nnvm/python/nnvm/to_relay.py

```diff
@@ -0,0 +1,530 @@
+    '__sub_scalar__': _to_subtract,
+    '__rsub_scalar__': _to_rsubtract,
+    'broadcast_sub': _to_subtract,
+    'elemwise_sub': _to_subtract,
+    # Multiply
+    '__mul_scalar__': _to_multiply,
+    'broadcast_mul': _to_multiply,
+    'elemwise_mul': _to_multiply,
+    # Division
+    '__div_scalar__': _to_divide,
+    'broadcast_div': _to_divide,
+    'elemwise_div': _to_divide,
+    # Comparsion
+    'greater': _to_greater,
+    'greater_equal': _to_greater_equal,
+    'less': _to_less,
+    'less_equal': _to_less_equal,
+    # Activations
+    'sigmoid': _to_sigmoid,
+    'clip': _to_clip,
+    'round': _to_round,
+    'cast': _to_cast,
+    'expand_dims': _to_expand_dims,
+    'broadcast_to': broadcast_to,
+    '__rshift_scalar__': _to_rshift,
+    'copy': _to_copy,
+    'global_avg_pool2d': _to_global_avg_pool2d,
+    'avg_pool2d': _to_avg_pool2d,
+    'conv2d_transpose': _to_conv2d_transpose,
+    'tanh': _to_tanh,
+    'upsampling': _to_upsampling,
+    'pad': _to_pad,
+    'exp': _to_exp,
+    'log': _to_log,
+    'leaky_relu': _to_leaky_relu,
+    'full_like': _to_full_like,
+    'negative': _to_negative,
+    'strided_slice': _to_strided_slice,
+    'split': _to_split
+}
+
+
+def to_relay(graph, shape_dict, dtype_dict, params):
+    """Convert an NNVM graph into the corresponding Relay expression.
+
+    Parameters
+    ----------
+    graph : Graph
+       The input graph.
+
+    shape_dict : dict of str to shape
+       The input shape.
+
+    dtype_dict : dict of str to shape
+       The input shape.
+
+    params : dict of str to array
+        The parameters.
+
+    Returns
+    -------
+    expr : relay.Expr
+        The corresponding Relay expression.
+    """
+    if isinstance(graph, Symbol):
+        graph = graph_create(graph)
+
+    param_shapes = dict((k, params[k].shape) for k in params)
+    shape_dict = shape_dict.copy()
+    shape_dict.update(param_shapes)
+    graph = graph_attr.set_shape_inputs(graph, shape_dict)
+    graph = graph_attr.set_dtype_inputs(graph, dtype_dict)
+    graph = graph.apply(["InferShape", "InferType"])
+    shape = graph.json_attr("shape")
+    dtype = [graph_attr.TCODE_TO_DTYPE[di] for di in graph.json_attr("dtype")]
+    heads = [x[0] for x in json.loads(graph.json())['heads']]
+
+    gidx = graph.index
+    relay_map = {}
+    params = []
+    output_ids = []
+
+    for nid, node in enumerate(gidx.nodes):
+        children = []
+        for i in node['inputs']:
+            child = relay_map[i[0]]
+            if isinstance(child, expr.TupleWrapper):
+                children.append(child[i[1]])
+            else:
+                children.append(child)
+
+        oshape = shape[gidx.entry_id(nid, 0)]
+        odtype = dtype[gidx.entry_id(nid, 0)]
+        attrs = node.get("attrs", {})
+        node_name = node["name"]
+        op_name = node["op"]
+
+        if op_name == "null":
+            v = var(node_name, shape=oshape, dtype=odtype)
+            params.append(v)
```

Comment:
This need to be updated to support the multiple output cases, consider directly reuse logic(we can refactor the function to take convert_map as argument) https://github.com/dmlc/tvm/blob/master/python/tvm/relay/frontend/mxnet.py#L469

Reply:
There is code below to handle multiple outputs, it will generate a tuple of all the outputs unless I'm missing something.

Reply:
I see that now, the two code handled it differently. In the mxnet converter code relay_map is asked to always record a list, tuple, or TuppleWrapper, and when we get the child we always fetch the corresponding index. This is going to be helpful for cases when we need post proc for multiple outputs(which can directly return a list).

The reason that I thought it only works with single output is due to the fact that when you get oshape, you only get output 0(instead of a list of all of them), so for functions like split, you only get the first output